### DISCUSSION TITLE
Does AOT + IL removal really work with BlazorWASM?

### DISCUSSION THREAD

**Discussion by jordi-z80** [dotnet/runtime#105635] [Q&A] [answered]
Hi, I'm trying to do a Blazor Webassembly app on net 8. One of the requirements is compiling it with AOT and removing IL data. I'm following this: https://learn.microsoft.com/en-us/aspnet/core/blazor/webassembly-build-tools-and-aot?view=aspnetcore-8.0

Basically I create the project  

`dotnet new blazorwasm -o BlazorApp`

Edit the project to add the AOT bits

```
	<RunAOTCompilation>true</RunAOTCompilation>
	<WasmStripILAfterAOT>true</WasmStripILAfterAOT>
```

And then just publish

`dotnet publish -c Release`

It takes a while to compile and at the end I have files at \bin\Release\net8.0\publish

However, if I go to wwwroot\\_framework, and I wasm2wat the main wasm BlazorApp.wasm (or any other WASM, for that matter) it seems like the file still contains webcil. These are the content headers:

```
(module
  (type (;0;) (func (param i32)))
  (type (;1;) (func (param i32 i32)))
  (import "webcil" "memory" (memory (;0;) 1))
  (func (;0;) (type 0) (param $destPtr i32)
    local.get $destPtr
    i32.const 0
    i32.const 4
    memory.init 0)
  (func (;1;) (type 1) (param $d i32) (param $n i32)
    local.get $d
    i32.const 0
    local.get $n
    memory.init 1)
  (global (;0;) i32 (i32.const 0))
  (global (;1;) i32 (i32.const 0))
  (export "webcilVersion" (global 0))
  (export "webcilSize" (global 1))
  (export "getWebcilSize" (func 0))
  (export "getWebcilPayload" (func 1))
  (data (;0;) "LB\00\00")
  (data (;1;) "WbIL.......
```

Is this really native wasm, or is the webcil still being kept in the .wasm files?

**Comment by huoyaoyuan** [comment]
> One of the requirements is compiling it with AOT and removing IL data.

This is the goal of [NativeAOT-LLVM](https://github.com/dotnet/runtimelab/tree/feature/NativeAOT-LLVM), which is still in experiment.

**Comment by jordi-z80** [comment]
I don't understand. The microsoft article I linked says nothing about being experimental, and gives concrete instructions on how to AOT and stripe IL. Maybe I'm mixing two different concepts? (I'm quite new to wasm)

**Reply by huoyaoyuan** [reply]
I'm not experienced in wasm either. It seems that my knowledge is outdated.

**Reply by teo-tsirpanis** [reply, marked as answer]
The WASM AOT in the article you linked is a different technology from NativeAOT-LLVM and supported since .NET 6, and cannot always completely remove IL.

**Reply by jordi-z80** [reply]
Thanks for the info, I'll try with NativeaOT-LLVM, it seems like the Microsoft document is quite misleading.
